File: JavaScript/059 Spiral Matrix II.js
// Given a positive integer n, generate a square matrix filled with elements from 1 to n^2 in spiral order.

// Example:

// Input: 3
// Output:
// [
//  [ 1, 2, 3 ],
//  [ 8, 9, 4 ],
//  [ 7, 6, 5 ]
// ]


/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function (n) {
    matrix = [];
    for (let i = 0; i < n; i++) {
        matrix.push([]);
    }

    let num = 0;
    for (let i = 0; i <= Math.floor((n - 1) / 2); i++) {
        let x = i;
        let y = i;
        for (; x < n - i; x++) {
            matrix[y][x] = ++num;
        }
        x--;
        y++;
        for (; y < n - i; y++) {
            matrix[y][x] = ++num;
        }
        y--;
        x--;
        for (; x >= i; x--) {
            matrix[y][x] = ++num;
        }
        x++;
        y--;
        for (; y > i; y--) {
            matrix[y][x] = ++num;
        }
    }
    return matrix;
};

// test
console.log(generateMatrix(3));

File: Python/059 Spiral Matrix II.py
# Given a positive integer n, generate a square matrix filled with elements from 1 to n^2 in spiral order.

# Example:

# Input: 3
# Output:
# [
#  [ 1, 2, 3 ],
#  [ 8, 9, 4 ],
#  [ 7, 6, 5 ]
# ]


class Solution(object):
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        matrix = [[0] * n for i in range(0, n)]

        num = 0
        for i in range(0, (n - 1) // 2 + 1):
            x, y = i, i
            while x < n - i:
                num += 1
                matrix[y][x] = num
                x += 1
            x -= 1
            y += 1
            while y < n - i:
                num += 1
                matrix[y][x] = num
                y += 1
            y -= 1
            x -= 1
            while x >= i:
                num += 1
                matrix[y][x] = num
                x -= 1
            x += 1
            y -= 1
            while y > i:
                num += 1
                matrix[y][x] = num
                y -= 1
        return matrix

# test
s = Solution()
print(s.generateMatrix(4))
